From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Mon, 8 Jan 2024 18:52:59 +0100
Subject: [PATCH] Modifiable Bukkit enums - Complete legacy conversion - Add
 missing conversions

Makes sure that every legacy `Material` can be converted to a non-legacy non-`AIR` (except if the original is `LEGACY_AIR`) `Material`.

Specifically, fixes the following legacy conversions:
* Skulls
* Long grass and its variants

License: GPL-3.0 (https://www.gnu.org/licenses/gpl-3.0.html)
Fiddle - https://fiddlemc.org

diff --git a/src/main/java/org/bukkit/craftbukkit/legacy/CraftLegacy.java b/src/main/java/org/bukkit/craftbukkit/legacy/CraftLegacy.java
index d9f731d605add107c0235c8f8db009bb924fb8a2..bd3a4db2db5232a28f43c9ea3ce5446960b0c5ce 100644
--- a/src/main/java/org/bukkit/craftbukkit/legacy/CraftLegacy.java
+++ b/src/main/java/org/bukkit/craftbukkit/legacy/CraftLegacy.java
@@ -424,6 +424,67 @@ public final class CraftLegacy {
                 itemToMaterial.put(newMaterial, matData);
             }
         }
+
+        // Fiddle start - modifiable Bukkit enums - Material - make legacy conversion complete - add LEGACY_SKULL
+        for (byte data = 0; data < 16; data++) {
+            var materialData = new MaterialData(Material.LEGACY_SKULL, data);
+            Item nonLegacyItem;
+            Block nonLegacyBlock;
+            switch (data) {
+                case 1 -> {
+                    nonLegacyItem = Items.WITHER_SKELETON_SKULL;
+                    nonLegacyBlock = Blocks.WITHER_SKELETON_SKULL;
+                }
+                case 2 -> {
+                    nonLegacyItem = Items.ZOMBIE_HEAD;
+                    nonLegacyBlock = Blocks.ZOMBIE_HEAD;
+                }
+                case 3 -> {
+                    nonLegacyItem = Items.PLAYER_HEAD;
+                    nonLegacyBlock = Blocks.PLAYER_HEAD;
+                }
+                case 4 -> {
+                    nonLegacyItem = Items.CREEPER_HEAD;
+                    nonLegacyBlock = Blocks.CREEPER_HEAD;
+                }
+                case 5 -> {
+                    nonLegacyItem = Items.DRAGON_HEAD;
+                    nonLegacyBlock = Blocks.DRAGON_HEAD;
+                }
+                default -> {
+                    nonLegacyItem = Items.SKELETON_SKULL;
+                    nonLegacyBlock = Blocks.SKELETON_SKULL;
+                }
+            }
+            materialToItem.put(materialData, nonLegacyItem);
+            materialToBlock.put(materialData, nonLegacyBlock);
+        }
+        // Fiddle end - modifiable Bukkit enums - Material - make legacy conversion complete - add LEGACY_SKULL
+
+        // Fiddle start - modifiable Bukkit enums - Material - make legacy conversion complete - ensure correct LEGACY_LONG_GRASS
+        for (byte data = 0; data < 16; data++) {
+            var materialData = new MaterialData(Material.LEGACY_LONG_GRASS, data);
+            Item nonLegacyItem;
+            Block nonLegacyBlock;
+            switch (data) {
+                case 1 -> {
+                    nonLegacyItem = Items.GRASS;
+                    nonLegacyBlock = Blocks.GRASS;
+                }
+                case 2 -> {
+                    nonLegacyItem = Items.FERN;
+                    nonLegacyBlock = Blocks.FERN;
+                }
+                default -> {
+                    nonLegacyItem = Items.DEAD_BUSH;
+                    nonLegacyBlock = Blocks.DEAD_BUSH;
+                }
+            }
+            materialToItem.put(materialData, nonLegacyItem);
+            materialToBlock.put(materialData, nonLegacyBlock);
+        }
+        // Fiddle end - modifiable Bukkit enums - Material - make legacy conversion complete - ensure correct LEGACY_LONG_GRASS
+
     }
 
     public static void main(String[] args) {
